### ISSUE TITLE
Backprapogation Algorithm

### ISSUE THREAD

**Issue by YogeshShitole** [Boyiliee/AOD-Net#4]
Hi @Boyiliee ,
Can you please tell what backprapogation algo. you used for gradient calculation
I went through your paper but did not find any mention of backprapogation algo./function which you
used for training your AOD model although other training parameters such as learning rate, momentum, weight decay etc...  are mentioned

**Comment by Boyiliee** [comment]
Thanks for your questions. I used SGD as backpropogation algo. Hope this helps.

**Comment by YogeshShitole** [comment]
Hi @Boyiliee 
Thanks for quick response...
"Loss is not converging during training"
I made the changes in train script but during training loss is not converging 
I am training network with NYU2 database with 27, 256 as mentioned in your paper
I have used your test_template.prototxt as training.prototxt   and I am training network with 150000 iteration ((~27,567/batchsize=8)*40 epoch)
as Euclidean loss layer is used in prototxt loss is not converging over the period it remain same around 16000 
upon training completion I even checked for inference with trained model_iter_150000.caffemodel instead of your AOD_Net.caffemodel but it is producing completely white image as output instead of giving dehazed image

below is my solver.prototxt content
-------------------------------------------------
net:"Training.prototxt"
# lr for fine-tuning should be lower than when starting from scratch
base_lr: 0.001
lr_policy: "fixed"
display: 20
max_iter: 150000
momentum: 0.9
weight_decay: 0.0001
snapshot: 15000
snapshot_prefix: "models/model"
# uncomment the following to default to CPU mode solving
solver_mode: GPU
type: "SGD"

# training.prototxt is same as test_template.prototxt with only modification for data/label layer
# input_dim: 1 -----> input_dim: {batchSize} 

My train script
-----------------------------------------------
import os
import numpy as np
from pylab import *
import re
import random
import cv2
print cv2.__version__
import ntpath
ntpath.basename('a/b/c')

def path_leaf(path):
    head, tail = ntpath.split(path)
    return tail or ntpath.basename(head)

Train_DIR = '../data/AODtrain/training/'
Label_DIR = '../data/AODtrain/original/'

# Network Training parameters for input Image data
height = 480
width = 640
batch =  8 # batch size

import sys
sys.path.append("/home/ubuntu//Tools/caffe/python/")
import caffe

def EditFcnProto(templateFile, height, width, batch_size):
	with open(templateFile, 'r') as ft:
		template = ft.read()
        # print templateFile
        outFile = 'Training.prototxt'
        with open(outFile, 'w') as fd:
            fd.write(template.format(height=height,width=width,batchSize=batch_size))

def createBatch(img_dir, label_dir, batch_size):
    batchdata = []
    labelbatchdata = []
    for i in range(batch_size):
        fname = random.choice(os.listdir(img_dir))
        imagepath = Train_DIR + fname
        # print fname
        # print imagepath
        labelpath = label_dir + fname.split('_')[0] + '_' + fname.split('_')[1] + '.jpg'
        # print labelpath

        npstore = caffe.io.load_image(imagepath)
        labelstore = caffe.io.load_image(labelpath)

        data = npstore
        data = data.transpose((2, 0, 1))
        batchdata.append(data)

        label = labelstore
        label = label.transpose((2, 0, 1))
        labelbatchdata.append(label)

    return batchdata, labelbatchdata


def train():
    caffe.set_mode_gpu()
    caffe.set_device(0)

    # same as test_template .prototxt only modification input_dim: 1 -----> input_dim: {batchSize}
    templateFile = 'train_template.prototxt'  
    EditFcnProto(templateFile, height, width, batch)

    solver = caffe.SGDSolver('solver.prototxt')
    # solver = caffe.get_solver('solver.prototxt')


    niter = 150000
    train_loss = np.zeros(niter)

    f = open('loss.txt', 'w')

    for it in range(niter):
        batchdata, labelbatchdata = createBatch(Train_DIR, Label_DIR, batch)
        solver.net.blobs['data'].data[...] = batchdata;
        solver.net.blobs['label'].data[...] = labelbatchdata;
        solver.step(1)
        train_loss[it] = solver.net.blobs['loss'].data
        f.write('{0: d} '.format(it))
        f.write('{0: f}\n'.format(train_loss[it]))

    f.close()


if __name__ == '__main__':
    train()

I am confused what is going wrong here, please tell me what I am doing wrong and suggest how to proceed
basically I am trying to reproduce your paper for learning

Thanks
